### ISSUE TITLE
Widening API mismatches flow condition and (previous) values

### ISSUE THREAD

**Issue by ymand** [llvm/llvm-project#76462]
The [widening API requires](https://github.com/llvm/llvm-project/blob/a01b58aef0e42fb1b52e358adf4c56678a884d37/clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h#L108-L134) that analyses return the previous value when it is equivalent to the current values's widening. But, that equivalence is *with respect to the previous environment (and its flow condition)*, while the widening operation will save the *current* environment. So, when checking equivalence, the operation succeeds, but then the resulting environment is incoherent -- the previous value is saved with the current flow condition. It is easy to trigger situations where the current flow condition is too weak to prove assertions about the previous value (for example, that a given property is false).

The API must provide a way for the widened value to be correct with respect to the saved environment.

**Comment by martinboehme** [comment]
Wow, this is subtle.

> It seems better to modify the API to allow implementations to independently return the desired widened value and the indication of whether or not it is identical to the previous value.

Agree, this seems preferable. Some comments:

*  Until we do this, I think we should at least document this pitfall in the API comments?
*  Even after we've made an API change, it seems it would be possible for the callee to return the previous value -- which would then cause the same issue. So we should probably still include some documentation saying not to do this (and stressing that the value will be interpreted with respect to the current environment)?

**Comment by ymand** [comment]
Agreed on both points -- we should document this issue as the first step, since (as you point out) it's going to matter even if we improve the API.
